File: ios/Sources/KlinikMedicationsFeature/PrescribingModel.swift
import Foundation
import KlinikAPI
import KlinikCore

public enum PrescribingPhase: Sendable, Equatable {
    case loading
    case loaded
    case empty
    case failed(String)
}

public struct PrescribingState: Sendable, Equatable {
    public var phase: PrescribingPhase = .loading
    public var medications: [MedicationView] = []
    /// What the reference knows about this patient's drugs together.
    public var interactions: InteractionCheck?
    public var busyId: String?
    public var error: String?
    /// Set after a write, so the sheet knows to close.
    public var lastWritten: MedicationView?

    public init() {}

    public var active: [MedicationView] {
        medications.filter { $0.medication.isActive }
    }

    public var awaitingApproval: [MedicationView] {
        medications.filter(\.medication.awaitingApproval)
    }

    public var stopped: [MedicationView] {
        medications.filter { $0.medication.stoppedAt != nil }
    }
}

/**
 * The clinician's side of the medication module (spec M9).
 *
 * Until this existed the doctor could read a plan and not write one, which is
 * the wrong way round: the spec has the clinician define the plan and the
 * patient tick the doses off.
 *
 * The interaction check is loaded beside the list rather than only after a
 * write, because a doctor opening this screen is often opening it *to* decide
 * whether to add something — and the reference is explicit that a missing
 * warning is not a safety claim.
 */
@MainActor
public final class PrescribingModel {
    private let api: MedicationsAPI
    private let patientId: String

    private var state = PrescribingState()

    public init(api: MedicationsAPI, patientId: String) {
        self.api = api
        self.patientId = patientId
    }

    public func currentState() -> PrescribingState { state }

    public func load() async {
        do {
            async let medications = api.forPatient(patientId)
            // Best-effort: the reference table being unavailable must not stop
            // a doctor reading the plan.
            async let interactions = try? await api.interactions(patientId: patientId)

            state.medications = try await medications
            state.interactions = await interactions
            state.phase = state.medications.isEmpty ? .empty : .loaded
        } catch let error as APIError {
            state.phase = .failed(L10n.message(for: error))
        } catch {
            state.phase = .failed(L10n.string("error.server"))
        }
    }

    public func prescribe(_ prescription: Prescription) async -> Bool {
        state.error = nil
        state.busyId = "new"

        defer { state.busyId = nil }

        do {
            let written = try await api.prescribe(patientId: patientId, prescription)
            state.lastWritten = written
            await load()

            return true
        } catch let error as APIError {
            state.error = L10n.message(for: error)
        } catch {
            state.error = L10n.string("error.server")
        }

        return false
    }

    public func approve(_ medicationId: String) async {
        await write(medicationId) {
            try await self.api.approve(patientId: self.patientId, medicationId: medicationId)
        }
    }

    public func stop(_ medicationId: String) async {
        await write(medicationId) {
            try await self.api.stop(patientId: self.patientId, medicationId: medicationId)
        }
    }

    private func write(_ id: String, _ work: () async throws -> MedicationView) async {
        state.busyId = id
        state.error = nil

        defer { state.busyId = nil }

        do {
            let updated = try await work()

            // Replaced with what the server returned rather than flipped
            // locally: an approval the server refused must not look applied.
            if let index = state.medications.firstIndex(where: { $0.id == id }) {
                state.medications[index] = updated
            }

            // A stop or an approval changes what the interaction check compares.
            state.interactions = try? await api.interactions(patientId: patientId)
        } catch let error as APIError {
            state.error = L10n.message(for: error)
        } catch {
            state.error = L10n.string("error.server")
        }
    }
}

File: android/feature/medications/src/main/kotlin/xyz/klinik/feature/medications/PrescribingModel.kt
package xyz.klinik.feature.medications

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import xyz.klinik.network.ApiError
import xyz.klinik.network.InteractionCheck
import xyz.klinik.network.MedicationView
import xyz.klinik.network.MedicationsApi
import xyz.klinik.network.Prescription
import xyz.klinik.network.UiText
import xyz.klinik.network.uiText

sealed interface PrescribingPhase {
    data object Loading : PrescribingPhase
    data object Loaded : PrescribingPhase

    /** Nothing prescribed and nothing reported. Not a failure. */
    data object Empty : PrescribingPhase
    data class Failed(val message: UiText) : PrescribingPhase
}

data class PrescribingState(
    val phase: PrescribingPhase = PrescribingPhase.Loading,
    val medications: List<MedicationView> = emptyList(),
    /** What the reference knows about this patient's drugs together. */
    val interactions: InteractionCheck? = null,
    val busyId: String? = null,
    val error: UiText? = null,
) {
    val active: List<MedicationView> get() = medications.filter { it.medication.isActive }

    /**
     * What the patient says they are taking, waiting on a clinician.
     *
     * Its own list because it is a different question: these generate no doses
     * and count towards no adherence until somebody approves them, and a
     * clinician scanning one merged list would read them as prescribed.
     */
    val awaitingApproval: List<MedicationView>
        get() = medications.filter { it.medication.awaitingApproval }

    val stopped: List<MedicationView>
        get() = medications.filter { it.medication.stoppedAt != null }
}

/**
 * The clinician's side of the medication module (spec M9).
 *
 * Until this existed on Android a doctor could read a plan and not write one,
 * which is the wrong way round: the spec has the clinician define the plan and
 * the patient tick the doses off.
 *
 * The interaction check is loaded beside the list rather than only after a
 * write, because a doctor opening this screen is often opening it *to* decide
 * whether to add something. It is best-effort on its own: the reference table
 * being unavailable must not stop somebody reading the plan — and the check's
 * own contract is explicit that a missing warning is not a safety claim.
 */
class PrescribingModel(
    private val api: MedicationsApi,
    private val patientId: String,
) {
    private val _state = MutableStateFlow(PrescribingState())
    val state: StateFlow<PrescribingState> = _state.asStateFlow()

    suspend fun load() {
        _state.value = _state.value.copy(phase = PrescribingPhase.Loading)

        try {
            val medications = api.forPatient(patientId)
            val interactions = runCatching { api.interactions(patientId) }.getOrNull()

            _state.value = _state.value.copy(
                phase = if (medications.isEmpty()) {
                    PrescribingPhase.Empty
                } else {
                    PrescribingPhase.Loaded
                },
                medications = medications,
                interactions = interactions,
            )
        } catch (error: Throwable) {
            _state.value = _state.value.copy(phase = PrescribingPhase.Failed(messageFor(error)))
        }
    }

    suspend fun prescribe(prescription: Prescription): Boolean {
        _state.value = _state.value.copy(busyId = NEW, error = null)

        return try {
            api.prescribe(patientId, prescription)
            // Reloaded rather than appended: the server generates the doses and
            // recomputes the interaction check, and a locally inserted row
            // would show a schedule nobody had confirmed.
            load()
            _state.value = _state.value.copy(busyId = null)

            true
        } catch (error: Throwable) {
            _state.value = _state.value.copy(busyId = null, error = messageFor(error))
            false
        }
    }

    suspend fun approve(medicationId: String) = write(medicationId) { api.approve(medicationId) }

    suspend fun stop(medicationId: String) = write(medicationId) { api.stop(medicationId) }

    private suspend fun write(id: String, work: suspend () -> MedicationView): Boolean {
        _state.value = _state.value.copy(busyId = id, error = null)

        return try {
            val updated = work()

            _state.value = _state.value.copy(
                // Replaced with what the server returned rather than flipped
                // locally: an approval the server refused must not look
                // applied.
                medications = _state.value.medications.map {
                    if (it.medication.id == id) updated else it
                },
                // A stop or an approval changes what the check compares.
                interactions = runCatching { api.interactions(patientId) }.getOrNull()
                    ?: _state.value.interactions,
                busyId = null,
            )

            true
        } catch (error: Throwable) {
            _state.value = _state.value.copy(busyId = null, error = messageFor(error))
            false
        }
    }

    private fun messageFor(error: Throwable): UiText =
        (error as? ApiError)?.uiText() ?: UiText.Key("error.server")

    private companion object {
        /** The id a not-yet-written prescription is busy under. */
        const val NEW = "new"
    }
}
